src/sqlfluff/core/linter.py

```diff
@@ -152,15 +160,90 @@ def get_violations(
+        return violations
+
+    @staticmethod
+    def _ignore_masked_violations_single_line(
+        violations: List[SQLBaseError], ignore_mask: List[NoQaDirective]
+    ):
+        """Returns whether to ignore error for line-specific directives.
+
+        The "ignore" list is assumed to ONLY contain NoQaDirectives with
+        action=None.
+        """
+        for ignore in ignore_mask:
+            violations = [
+                v
+                for v in violations
+                if not (
+                    v.line_no() == ignore.line_no
+                    and (ignore.rules is None or v.rule_code() in ignore.rules)
+                )
+            ]
+        return violations
+
+    @staticmethod
+    def _should_ignore_violation_line_range(
+        line_no: int, ignore_rule: List[NoQaDirective]
+    ):
+        """Returns whether to ignore a violation at line_no."""
+        # Loop through the NoQaDirectives to find the state of things at
+        # line_no. Assumptions about "ignore_rule":
+        # - Contains directives for only ONE RULE, i.e. the rule that was
+        #   violated at line_no
+        # - Sorted in ascending order by line number
+        disable = False
+        for ignore in ignore_rule:
+            if ignore.line_no > line_no:
+                break
+            disable = ignore.action == "disable"
+        return disable
+
+    @classmethod
+    def _ignore_masked_violations_line_range(
+        cls, violations: List[SQLBaseError], ignore_mask: List[NoQaDirective]
+    ):
+        """Returns whether to ignore error for line-range directives.
+
+        The "ignore" list is assumed to ONLY contain NoQaDirectives where
+        action is "enable" or "disable".
+        """
+        result = []
+        for v in violations:
+            # Find the directives that affect the violated rule "v", either
+            # because they specifically reference it or because they don't
+            # specify a list of rules, thus affecting ALL rules.
+            ignore_rule = sorted(
+                [
+                    ignore
+                    for ignore in ignore_mask
+                    if not ignore.rules
+                    or (v.rule_code() in cast(Tuple[str, ...], ignore.rules))
+                ],
+                key=lambda ignore: ignore.line_no,
+            )
+            # Determine whether to ignore the violation, based on the relevant
+            # enable/disable directives.
+            if not cls._should_ignore_violation_line_range(v.line_no(), ignore_rule):
+                result.append(v)
+        return result
+
+    def _ignore_masked_violations(
+        self, violations: List[SQLBaseError]
+    ) -> List[SQLBaseError]:
+        """Remove any violations specified by ignore_mask.
+
+        This involves two steps:
+        1. Filter out violations affected by single-line "noqa" directives.
+        2. Filter out violations affected by disable/enable "noqa" directives.
+        """
```

Comment:
Just want to call out, really love that you did this in a way that allowed both of these to co-exist. Not only do I think there is a place for both of them, I love that we're not breaking anyone's code.

Reply:
Thanks! TBH, I followed the pylint syntax pretty closely, so I'm not taking credit for this. 😊 I thought about stealing their code, but I noticed it uses GPL license, so I didn't even look at their code, just wrote it from scratch to avoid being evil.